src/cascadia/TerminalCore/TerminalSelection.cpp

```diff
@@ -478,3 +478,220 @@ void Terminal::ColorSelection(const COORD, const COORD, const TextAttribute)
+    THROW_IF_FAILED(ShortSub(positionWithOffsets.Y, gsl::narrow<SHORT>(_ViewStartIndex()), &positionWithOffsets.Y));
+    if (target == SelectionTarget::Start)
+    {
+        _selectionAnchor = positionWithOffsets;
+    }
+    else
+    {
+        _endSelectionPosition = positionWithOffsets;
+    }
+    _selectionVerticalOffset = gsl::narrow<SHORT>(_ViewStartIndex());
+}
+
+// Method Description:
+// - update the endSelectionPosition anchor by one cell according to the direction provided.
+// Arguments:
+// - dir: the direction that the selection anchor will attempt to move to
+// Return Value:
+// - <none>
+void Terminal::_UpdateAnchorByCell(Direction dir, COORD& anchor)
+{
+    auto bufferViewport = _buffer->GetSize();
+    switch (dir)
+    {
+    case Direction::Left:
+        bufferViewport.DecrementInBounds(anchor);
+        break;
+    case Direction::Right:
+        bufferViewport.IncrementInBounds(anchor);
+        break;
+    case Direction::Up:
+        THROW_IF_FAILED(ShortSub(anchor.Y, 1, &anchor.Y));
+        break;
+    case Direction::Down:
+        THROW_IF_FAILED(ShortAdd(anchor.Y, 1, &anchor.Y));
+        break;
+    default:
+        // direction is not supported.
+        THROW_HR(E_NOTIMPL);
+        return;
+    }
+}
+
+// Method Description:
+// - update the endSelectionPosition anchor by one word according to the direction provided.
+// Arguments:
+// - dir: the direction that the selection anchor will attempt to move to
+// Return Value:
+// - <none>
+void Terminal::_UpdateAnchorByWord(Direction dir, COORD& anchor, SelectionTarget target)
+{
+    // we need this to be able to compare it later
+    // NOTE: if target = START --> return END; else return START
+    COORD otherSelectionAnchor = ((target == SelectionTarget::Start) ? _endSelectionPosition : _selectionAnchor);
+    otherSelectionAnchor.Y += _selectionVerticalOffset;
+
+    auto bufferViewport = _buffer->GetSize();
+    auto comparison = bufferViewport.CompareInBounds(anchor, otherSelectionAnchor);
+    switch (dir)
+    {
+    case Direction::Up:
+        THROW_IF_FAILED(ShortSub(anchor.Y, 1, &anchor.Y));
+        bufferViewport.Clamp(anchor);
+    case Direction::Left:
+        if (comparison < 0)
+        {
+            // get on new run, before expanding left
+            bufferViewport.DecrementInBounds(anchor);
+        }
+        anchor = _ExpandDoubleClickSelectionLeft(anchor);
+        if (comparison > 0)
+        {
+            // get off of run, after expanding left
+            bufferViewport.DecrementInBounds(anchor);
+        }
+        break;
+    case Direction::Down:
+        THROW_IF_FAILED(ShortAdd(anchor.Y, 1, &anchor.Y));
+        bufferViewport.Clamp(anchor);
+    case Direction::Right:
+        if (comparison > 0)
+        {
+            // get on new run, before expanding left
+            bufferViewport.IncrementInBounds(anchor);
+        }
+        anchor = _ExpandDoubleClickSelectionRight(anchor);

```

Comment:
Now that there are more ways to expand by a word than just double-clicking, should `_ExpandDoubleClickSelectionRight` and `_ExpandDoubleClickSelectionLeft` be renamed to something more appropriate?